llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp

```diff
@@ -167,6 +168,50 @@ AMDGPURegBankLegalizeCombiner::tryMatchRALFromUnmerge(Register Src) {
   return {nullptr, -1};
 }
 
+// Vgpr0, Vgpr1, ... = G_UNMERGE_VALUES UnmergeSrc (or defined elsewhere)
+// Sgpr0 = G_AMDGPU_READANYLANE Vgpr0
+// Sgpr1 = G_AMDGPU_READANYLANE Vgpr1
+// ...
+// Merge = G_MERGE_VALUES/G_BUILD_VECTOR/G_CONCAT_VECTORS Sgpr0, Sgpr1, ...
+// ->
+// If Vgpr sources are all from same G_UNMERGE_VALUES with sequential indices:
+//   return UnmergeSrc directly
+// Otherwise:
+//   build G_MERGE_VALUES/G_BUILD_VECTOR/G_CONCAT_VECTORS Vgpr0, Vgpr1, ...
+Register AMDGPURegBankLegalizeCombiner::tryFoldRALsIntoMergeLikeInstr(
+    GMergeLikeInstr &Merge) {
+  unsigned NumElts = Merge.getNumSources();
+  SmallVector<Register, 4> VgprSrcs;
+
+  // Collect VGPR sources from READANYLANEs.
+  for (unsigned i = 0; i < NumElts; ++i) {
+    auto [RAL, VgprSrc] =
+        tryMatch(Merge.getSourceReg(i), AMDGPU::G_AMDGPU_READANYLANE);
+    if (!RAL)
+      return {};
+    VgprSrcs.push_back(VgprSrc);
+  }
+
+  // Check if all from same unmerge with sequential indices.
+  auto [Unmerge, Idx] = tryMatchRALFromUnmerge(Merge.getSourceReg(0));
+  if (Unmerge && Unmerge->getNumDefs() == NumElts && Idx == 0) {
+    bool AllFromSameUnmerge = true;
+    for (unsigned i = 1; i < NumElts; ++i) {
+      auto [UnmergeI, IdxI] = tryMatchRALFromUnmerge(Merge.getSourceReg(i));
+      if (UnmergeI != Unmerge || (unsigned)IdxI != i) {
+        AllFromSameUnmerge = false;
+        break;
+      }
+    }
+    if (AllFromSameUnmerge)
+      return Unmerge->getSourceReg();
+  }
```

Comment:
you already matched read-any-lanes, need to check if  VgprSrcs are all defined by unmerge
```suggestion

Register tryMatchUnmergeDefs(SmallVectorImpl<Register> &DefRegs,
                             MachineRegisterInfo &MRI) {
  auto *UnMerge = getOpcodeDef<GUnmerge>(DefRegs[0], MRI);
  if (!UnMerge || UnMerge->getNumDefs() != DefRegs.size())
    return {};
  for (unsigned i = 1; i < DefRegs.size(); ++i) {
    if (UnMerge->getReg(i) != DefRegs[i])
      return {};
  }
  return UnMerge->getSourceReg();
}

  if (Register UnmergeSrc = tryMatchUnmergeDefs(VgprSrcs, MRI))
    return UnmergeSrc;
```